src/libsyntax/ast.rs

```diff
@@ -1857,7 +1866,7 @@ pub struct Variant_ {
     pub attrs: Vec<Attribute>,
     pub data: VariantData,
     /// Explicit discriminant, e.g. `Foo = 1`
-    pub disr_expr: Option<P<Expr>>,
+    pub disr_expr: Option<Const>,
```

Comment:
It's a bit strange that `const` and `static` initializers are not treated as `Const` and not visited by `fn visit_const`, for example, despite being same "constant single-expression bodies".

Reply:
Is it *necessary* to introduce `Const` in AST, btw?
New node ids can be assigned during lowering to HIR too.

(I see that some stuff in resolve becomes shorted, but I haven't looked in detail yet.)

Reply:
Those have items (or trait/impl associated items) they're directly attached to, so they don't need any changes, but also changing them would complicate a bunch of logic that's straight-forward today and would have to jump through an extra hoop.

There's one potential exception to this, which is that `static` could be considered to be "more disjoint" from its initializer than `const` and separating them could allow better bookkeeping of the "allocation" vs "value the allocation is initialized with" (cc @oli-obk), but even then it seems a bit stretched.

I wanted to avoid calling the type `EmbeddedConst` because it's long and potentially even more confusing, but if anyone has a better suggestion, it could use being more explicit.

Reply:
the static vs its value won't get easier by changing this. Miri's bookkeeping about that is completely sane now.

Reply:
> Is it necessary to introduce `Const` in AST, btw?
> New node ids can be assigned during lowering to HIR too.

Sadly, yes, because the `DefId` tree is built from the AST, and nodes inside a `Const` need to be parented by it in that tree (this is where some of the ICEs for closures came from).

Reply:
So, what criterion should a constant `Expr` satisfy to become `Const` then? No direct parent with `DefId`?
Discriminants in enums have a direct parent with `DefId` to which they are attached - the variant itself, in this sense I don't see how they are different from `const`s or `static`s. Is preferring `Const` just a matter of convenience in this case?
Could all this be documented in a comment for `Const`?

Reply:
Also, expressions in range patterns have very similar properties to array sizes, etc - `THIS ... AND_THIS`, but I guess they don't need a body because you can't smuggle arbitrary expressions into them, paths at most.

Reply:
The criterion is a requirement of separate const-evaluation (which requires separate type-checking).

Array type lengths and enum discriminants don't have a parent body so they need a separate `DefId` and body to even be type-checked (array type lengths could have a `DefId` attached to the `NodeId` of the `TyArray`, but enum variants already have a `DefId` with a different meaning).

The count of an `ExprRepeat` is technically guaranteed to be nested inside another body, but type-checking the `ExprRepeat` requires const-evaluating the count, which currently can't be done without the count being a separate body.

In contrast, `const` and `static`

> Also, expressions in range patterns have very similar properties to array sizes

It might be a good idea to keep those separate too, but type-checking them could require inference from the outer body (because of associated consts), so it's tricky.

I think long-term we should perhaps keep such constant expressions (including ones from paths involving const generics, in the future), in a separate body *but* type-check them like closures (sharing the inference context of their parent body), while retaining the ability to const-eval them.

Reply:
Oh. will this make the difference between a variant's discriminant and it's isize value clearer?

Reply:
> will this make the difference between a variant's discriminant and it's isize value clearer?

Doubful.